packages/bp/src/core/dialog/dialog-engine.ts

```diff
@@ -219,79 +269,94 @@ export class DialogEngine {
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, timeoutNodeName)
       }
     }
 
     // Check for a timeout.flow.json and get the start node
     if (!timeoutNode) {
-      timeoutFlow = findFlowWithoutError('timeout.flow.json')
+      timeoutFlow = this.findFlowWithoutError(botId, 'timeout.flow.json')
       if (timeoutFlow) {
         const startNodeName = timeoutFlow.startNode
-        timeoutNode = findNodeWithoutError(timeoutFlow, startNodeName)
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, startNodeName)
       }
     }
 
     if (!timeoutNode || !timeoutFlow) {
       throw new TimeoutNodeNotFound(`Could not find any timeout node or flow for session "${sessionId}"`)
     }
 
-    if (event.state.context) {
-      event.state.context.currentNode = timeoutNode.name
-      event.state.context.currentFlow = timeoutFlow.name
-      event.state.context.queue = undefined
-      event.state.context.previousFlow = currentFlow.name
-      event.state.context.previousNode = currentNode?.name as string
-      event.state.context.jumpPoints = [
-        ...(event.state.context?.jumpPoints || []),
-        {
-          flow: currentFlow.name,
-          node: currentNode?.name as string
-        }
-      ]
-      event.state.context.hasJumped = true
-    }
+    this.fillContextForTransition(event, { currentFlow, currentNode, nextFlow: timeoutFlow, nextNode: timeoutNode })
 
     return this.processEvent(sessionId, event)
   }
 
-  private _endFlow(event: IO.IncomingEvent) {
-    event.state.context = {}
-    event.state.temp = {}
+  public async processConversationEnd(botId: string, sessionId: string, event: IO.IncomingEvent): Promise<boolean> {
+    this._debug(event.botId, event.target, 'processing conversation end')
+
+    await this._loadFlows(botId)
+
+    const currentFlow = this._findFlow(botId, event.state.context?.currentFlow!)
+    const currentNode = this.findNodeWithoutError(botId, currentFlow, event.state.context?.currentNode)
+
+    // Check for a conversation_end.flow.json
+    const conversationEndFlow = this.findFlowWithoutError(botId, 'conversation_end.flow.json')
+
+    if (currentFlow.name !== 'conversation_end.flow.json') {
+      const previousStackTraceSize = event.state.__stacktrace?.length
+      const api = await createForGlobalHooks()
+      await this.hookService.executeHook(new Hooks.BeforeConversationEnd(api, event))
+      // Check if stack trace size changed during hook execution, meaning that a transition was added
+      if (previousStackTraceSize !== event.state.__stacktrace?.length) {
+        await this.processEvent(sessionId, event)
+        return false
+      }
+    }
+
+    // Don`t trigger on timeout or if it is already at the conversation end flow
+    if (
+      !conversationEndFlow ||
+      currentFlow.name === 'conversation_end.flow.json' ||
+      currentFlow.name === 'timeout.flow.json'
+    ) {
+      return true
+    }
```

Comment:
In this case, I would still need to process the hook before, to check if there are any transitions, which happens to cancel the execution of the flow or the cleaning of state, which means a 50/50 split of the function, do you think this would this be worth it?

Reply:
I just think this function is bloated. and untestable